presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/file/FileHiveMetastore.java

```diff
@@ -1031,8 +1031,7 @@ private synchronized void setTablePrivileges(
 
             Path permissionsDirectory = getPermissionsDirectory(table);
 
-            metadataFileSystem.mkdirs(permissionsDirectory);
-            if (!metadataFileSystem.isDirectory(permissionsDirectory)) {
+            if (!metadataFileSystem.mkdirs(permissionsDirectory) && !metadataFileSystem.isDirectory(permissionsDirectory)) {
```

Comment:
Yes, we should check `mkdirs` result. But no, we should silently ignore `false` result from `mkdirs` _in general case_.

Reply:
That is what this is doing although the readability is a bit weird, so it's checking if it couldn't create the directories itself then it will check to make sure the path is already a directory. For readability could change to:
```
boolean created = metadataFileSystem.mkdirs(permissionsDirectory);
if (!created && !metadataFileSystem.isDirectory(permissionsDirectory)) {
    throw new PrestoException(HIVE_METASTORE_ERROR, "Could not create permissions directory");
}
```
or inverse the double negatives (although seems a bit harder to read actually):
```
boolean created = metadataFileSystem.mkdirs(permissionsDirectory);
if (!(created || metadataFileSystem.isDirectory(permissionsDirectory))) {
    throw new PrestoException(HIVE_METASTORE_ERROR, "Could not create permissions directory");
}
```
Could also reverse the ordering - bit of an implementation detail but this would then make S3 List have to get called where in the current code it just can short circuit on mkdirs always returning true.

Reply:
I just added the separate line to try to make this a bit more readable, cc @findepi